File: app/src/main/java/com/ourcookbook/domain/model/Ingredient.kt
package com.ourcookbook.domain.model

import java.util.UUID

/**
 * Domain model for Ingredient
 * Represents a single ingredient in a recipe
 * 
 * Contains all information about an ingredient including amount, unit, name,
 * and additional notes. Supports both metric and imperial measurement systems.
 */
data class Ingredient(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val amount: String? = null,
    val unit: String? = null,
    val notes: String? = null,
    val order: Int = 0
) {
    fun isValid(): Boolean {
        return name.isNotBlank()
    }
    
    // Display string for the ingredient
    val displayString: String get() {
        return buildString {
            if (!amount.isNullOrBlank()) {
                append(amount)
                if (!unit.isNullOrBlank()) {
                    append(" ").append(unit)
                }
                append(" ")
            }
            append(name)
            if (!notes.isNullOrBlank()) {
                append(" (").append(notes).append(")")
            }
        }.trim()
    }
    
    // Simple display without notes
    val simpleDisplay: String get() {
        return buildString {
            if (!amount.isNullOrBlank()) {
                append(amount)
                if (!unit.isNullOrBlank()) {
                    append(" ").append(unit)
                }
                append(" ")
            }
            append(name)
        }.trim()
    }
    
    companion object {
        fun create(
            name: String,
            amount: String? = null,
            unit: String? = null,
            notes: String? = null,
            order: Int = 0
        ): Ingredient {
            return Ingredient(
                name = name,
                amount = amount,
                unit = unit,
                notes = notes,
                order = order
            )
        }
    }
}

File: webapp/src/types/ingredient.ts
// Ingredient — a single ingredient in a recipe. Ported from Ingredient.kt.

export interface Ingredient {
  id: string
  name: string
  amount?: string | null
  unit?: string | null
  notes?: string | null
  order: number
}

export function createIngredient(
  name: string,
  amount: string | null = null,
  unit: string | null = null,
  notes: string | null = null,
  order = 0,
): Ingredient {
  return { id: crypto.randomUUID(), name, amount, unit, notes, order }
}

export function ingredientDisplay(ing: Ingredient): string {
  const parts: string[] = []
  if (ing.amount) {
    parts.push(ing.unit ? `${ing.amount} ${ing.unit}` : ing.amount)
  }
  parts.push(ing.name)
  const base = parts.join(' ')
  return ing.notes ? `${base} (${ing.notes})` : base
}
